test/evp_test.c

```diff
@@ -2288,30 +2346,75 @@ static int decapsulate(EVP_TEST *t, EVP_PKEY_CTX *ctx, const char *op,
     ret = 1;
 err:
     OPENSSL_free(out);
-    ctrl2params_free(params, params_n, params_n_allocated);
+    if (sk_OPENSSL_STRING_num(kdata->init_ctrls) > 0)
+        ctrl2params_free(params, params_n, params_n_allocated);
     return ret;
 }
 
 static int kem_test_run(EVP_TEST *t)
 {
-    int ret = 0;
+    int ret = 0, found_key = 0;
+    EVP_PKEY *pkey = NULL;
     KEM_DATA *kdata = t->data;
     unsigned char *wrapped = NULL, *secret = NULL;
 
+    /* Generate either public or private key based on given params */
+    if (kdata->keyname != NULL) {
+        /* Previously generated private key */
+        found_key = find_key(&pkey, kdata->keyname, private_keys);
+        if (found_key == 0 || pkey == NULL) {
+            TEST_info("skipping, key '%s' is disabled", kdata->keyname);
+            t->skip = 1;
+            goto ok;
+        }
+    } else if (kdata->encoded_pub_key != NULL) {
+        /* Encoded public key */
+        if ((pkey = EVP_PKEY_new_raw_public_key_ex(libctx, kdata->algname,
+                                                   propquery,
+                                                   kdata->encoded_pub_key,
+                                                   kdata->encoded_pub_key_len)) == NULL) {
+            t->err = "TEST_PARSE_PUBLIC_KEY_ERROR";
+            goto ok;
+        }
+    } else if (kdata->encoded_priv_key != NULL) {
+        /* Encoded private key */
+        if ((pkey = EVP_PKEY_new_raw_private_key_ex(libctx, kdata->algname,
+                                                    propquery,
+                                                    kdata->encoded_priv_key,
+                                                    kdata->encoded_priv_key_len)) == NULL) {
+            t->err = "TEST_PARSE_PRIVATE_KEY_ERROR";
+            goto ok;
+        }
+    } else {
+        TEST_info("Missing parameters to create key");
+        goto err;
+    }
+
+    if (!TEST_ptr(kdata->ctx = EVP_PKEY_CTX_new_from_pkey(libctx, pkey, propquery)))
+        goto err;
+
     if (kdata->input == NULL) {
         size_t wrappedlen = 0, secretlen = 0;
 
         ret = encapsulate(t, kdata->ctx, kdata->op, &wrapped, &wrappedlen,
                           &secret, &secretlen);
         if (ret == 0 || t->err != NULL)
             goto err;
-        ret = decapsulate(t, kdata->ctx, kdata->op, wrapped, wrappedlen,
-                          secret, secretlen);
+
+        /* Also attempt to decrypt if we have the private key */
+        if (found_key || kdata->encoded_priv_key != NULL)
```

Comment:
Wouldn't it be better for the test to fail if there's no private key available?

Reply:
The tests for ML-KEM-768 encap only have the encoded public key (and no private key)